### PR TITLE
update .21 core app insights sample with core bot 4.5.1 changes 

### PR DESCRIPTION
**Pull request by garypretty**
- Updates App Insights sample with the changes made to Core Bot in 4.5.1
- Set TelemetryClient property on MainDialog / BookingDialog in startup.cs as opposed to accepting via constructor
- Pass the TelemetryClient to the Luis Recognizer

### CODE REVIEW

**Comment by WashingtonKayaker** [comment]
Program.cs does not show up as one of the files that has changed, however if I compare this PR with sample 13.Core-Bot in the master branch, the 2 copyright lines in the comments at the top of the file are missing.

**Comment by garypretty** [comment]
@WashingtonKayaker good catch. I'll add it.

**Comment by garypretty** [comment]
@WashingtonKayaker ah yes. Great. I'll hopefully update tomorrow and let you know once done. 👍

File: samples/csharp_dotnetcore/21.corebot-app-insights/CoreBot-App-Insights.csproj [hunk 1]
```diff
@@ -14,4 +23,12 @@
     <PackageReference Include="Microsoft.Recognizers.Text.DataTypes.TimexExpression" Version="1.1.5" />
   </ItemGroup>
 
+  <ItemGroup>
```
Comment (gabog) [inline samples/csharp_dotnetcore/21.corebot-app-insights/CoreBot-App-Insights.csproj]: I don't think we need to copy the luis model to output, consider removing this itemgroup (check original corebot for details

**Comment by gabog** [comment]
> using Microsoft.ApplicationInsights.Extensibility;

Insert (C) hader

---
Refers to: samples/csharp_dotnetcore/21.corebot-app-insights/Startup.cs:1 in 9eddaca. [](commit_id = 9eddacaad99ad5b60df76331024816426615f115, deletion_comment = False)

File: samples/csharp_dotnetcore/21.corebot-app-insights/AdapterWithErrorHandler.cs [hunk 2]
```diff
@@ -6,14 +6,16 @@
...
 namespace Microsoft.BotBuilderSamples
 {
     public class AdapterWithErrorHandler : BotFrameworkHttpAdapter
     {
-        public AdapterWithErrorHandler(ICredentialProvider credentialProvider, ILogger<BotFrameworkHttpAdapter> logger, IMiddleware middleware, ConversationState conversationState = null)
-            : base(credentialProvider)
+        public AdapterWithErrorHandler(IConfiguration configuration, ILogger<BotFrameworkHttpAdapter> logger, IMiddleware middleware, ConversationState conversationState = null)
+            : base(configuration, logger)
         {
             Use(middleware);
```
Comment (gabog) [inline samples/csharp_dotnetcore/21.corebot-app-insights/AdapterWithErrorHandler.cs]: I know this is the way it was in the sample, but can we change IMiddleware middleware to TelemetryInitializerMiddleware telemetryInitializer? IMiddleware and the variable name is too broad and can mean anything.

File: samples/csharp_dotnetcore/21.corebot-app-insights/Startup.cs [hunk 3]
```diff
@@ -53,8 +50,31 @@ public void ConfigureServices(IServiceCollection services)
             // Create the Conversation state. (Used by the Dialog system itself.)
             services.AddSingleton<ConversationState>();
 
-            // The Dialog that will be run by the bot.
-            services.AddSingleton<MainDialog>();
+            // Register LUIS recognizer
+            services.AddSingleton<FlightBookingRecognizer>();
+
+            // Register the BookingDialog.
+            services.AddSingleton(sp =>
```
Comment (gabog) [inline samples/csharp_dotnetcore/21.corebot-app-insights/Startup.cs]: I personally prefer the DI way of injecting the IBotTelemetryClient (rather than the lamda)

File: samples/csharp_dotnetcore/21.corebot-app-insights/Startup.cs [hunk 4]
```diff
@@ -53,8 +50,31 @@ public void ConfigureServices(IServiceCollection services)
...
+            {
+                var telemetryClient = sp.GetService<IBotTelemetryClient>();
+                return new BookingDialog()
+                {
+                    TelemetryClient = telemetryClient
+                };
+            });
+
+            // The MainDialog that will be run by the bot passing and setting the TelemetryClient that will be used to log telemetry.
+            services.AddSingleton(sp =>
```
Comment (gabog) [inline samples/csharp_dotnetcore/21.corebot-app-insights/Startup.cs]: Same as above. I personally prefer the DI way of injecting the IBotTelemetryClient (rather than the lamda)

File: samples/csharp_dotnetcore/21.corebot-app-insights/Bots/DialogAndWelcomeBot.cs [hunk 5]
```diff
@@ -20,6 +20,7 @@ public DialogAndWelcomeBot(ConversationState conversationState, UserState userSt
         {
         }
 
+
```
[the remark this answers is not in the document]
Comment (garypretty) [reply samples/csharp_dotnetcore/21.corebot-app-insights/Bots/DialogAndWelcomeBot.cs]: Removed

File: samples/csharp_dotnetcore/21.corebot-app-insights/AdapterWithErrorHandler.cs [hunk 6]
```diff
@@ -6,14 +6,16 @@
 using Microsoft.Bot.Builder.Integration.AspNet.Core;
 using Microsoft.Bot.Builder.TraceExtensions;
 using Microsoft.Bot.Connector.Authentication;
```
[the remark this answers is not in the document]
Comment (garypretty) [reply samples/csharp_dotnetcore/21.corebot-app-insights/AdapterWithErrorHandler.cs]: Removed

File: samples/csharp_dotnetcore/21.corebot-app-insights/Dialogs/BookingDialog.cs [hunk 7]
```diff
@@ -5,27 +5,30 @@
...
+        private const string OriginStepMsgText = "Where are you traveling from?";
+
+        public BookingDialog()
             : base(nameof(BookingDialog))
         {
-            TelemetryClient = telemetryClient;
             AddDialog(new TextPrompt(nameof(TextPrompt))
             {
-                TelemetryClient = telemetryClient,
+                TelemetryClient = this.TelemetryClient
```
[the remark this answers is not in the document]
Comment (garypretty) [reply samples/csharp_dotnetcore/21.corebot-app-insights/Dialogs/BookingDialog.cs]: Removed instances of setting TelemetryClient on dialogs being added

File: samples/csharp_dotnetcore/21.corebot-app-insights/Dialogs/MainDialog.cs [hunk 8]
```diff
@@ -2,100 +2,154 @@
...
         {
-            Configuration = configuration;
+            _luisRecognizer = luisRecognizer;
             Logger = logger;
-            TelemetryClient = telemetryClient;
 
             AddDialog(new TextPrompt(nameof(TextPrompt))
             {
-                TelemetryClient = telemetryClient,
+                TelemetryClient = this.TelemetryClient
```
[the remark this answers is not in the document]
Comment (garypretty) [reply samples/csharp_dotnetcore/21.corebot-app-insights/Dialogs/MainDialog.cs]: Removed instances of setting TelemetryClient on dialogs being added

File: samples/csharp_dotnetcore/21.corebot-app-insights/Dialogs/BookingDialog.cs [hunk 9]
```diff
@@ -5,27 +5,30 @@
...
             });
             AddDialog(new ConfirmPrompt(nameof(ConfirmPrompt))
             {
-                TelemetryClient = telemetryClient,
+                TelemetryClient = this.TelemetryClient
```
[the remark this answers is not in the document]
Comment (garypretty) [reply samples/csharp_dotnetcore/21.corebot-app-insights/Dialogs/BookingDialog.cs]: Removed instances of setting TelemetryClient on dialogs being added

File: samples/csharp_dotnetcore/21.corebot-app-insights/Dialogs/BookingDialog.cs [hunk 10]
```diff
@@ -5,27 +5,30 @@
...
             });
             AddDialog(new DateResolverDialog()
             {
-                TelemetryClient = telemetryClient,
+                TelemetryClient = this.TelemetryClient
```
[the remark this answers is not in the document]
Comment (garypretty) [reply samples/csharp_dotnetcore/21.corebot-app-insights/Dialogs/BookingDialog.cs]: Removed instances of setting TelemetryClient on dialogs being added

File: samples/csharp_dotnetcore/21.corebot-app-insights/Dialogs/BookingDialog.cs [hunk 11]
```diff
@@ -36,7 +39,7 @@ public BookingDialog(IBotTelemetryClient telemetryClient)
                 FinalStepAsync,
             })
             {
-                TelemetryClient = telemetryClient,
+                TelemetryClient = this.TelemetryClient
```
[the remark this answers is not in the document]
Comment (garypretty) [reply samples/csharp_dotnetcore/21.corebot-app-insights/Dialogs/BookingDialog.cs]: Removed instances of setting TelemetryClient on dialogs being added

File: samples/csharp_dotnetcore/21.corebot-app-insights/CoreBot-App-Insights.csproj (hunk 1 above)
Comment (garypretty) [reply samples/csharp_dotnetcore/21.corebot-app-insights/CoreBot-App-Insights.csproj]: Agreed. Removed.

File: samples/csharp_dotnetcore/21.corebot-app-insights/AdapterWithErrorHandler.cs (hunk 2 above)
Comment (garypretty) [reply samples/csharp_dotnetcore/21.corebot-app-insights/AdapterWithErrorHandler.cs]: Agreed. Modified.

File: samples/csharp_dotnetcore/21.corebot-app-insights/Startup.cs (hunk 3 above)
Comment (garypretty) [reply samples/csharp_dotnetcore/21.corebot-app-insights/Startup.cs]: After review, I agree. I have reverted it to pass in the TelemetryClient into MainDialog via DI.  All other dialogs will have it set when they are added to the component dialog.
@WashingtonKayaker Please note this change.  I have also added a comment in MainDialog when dialogs are added to explicitly call out that their TelemetryClient property will be set automatically.

File: samples/csharp_dotnetcore/21.corebot-app-insights/Startup.cs (hunk 4 above)
Comment (garypretty) [reply samples/csharp_dotnetcore/21.corebot-app-insights/Startup.cs]: See above.

File: samples/csharp_dotnetcore/21.corebot-app-insights/Dialogs/DateResolverDialog.cs [hunk 12]
```diff
@@ -6,21 +6,31 @@
...
+        private const string PromptMsgText = "When would you like to travel?";
+        private const string RepromptMsgText = "I'm sorry, to make your booking please enter a full travel date including Day Month and Year.";
+
         public DateResolverDialog(string id = null)
             : base(id ?? nameof(DateResolverDialog))
         {
-            AddDialog(new DateTimePrompt(nameof(DateTimePrompt), DateTimePromptValidator));
+            AddDialog(new DateTimePrompt(nameof(DateTimePrompt), DateTimePromptValidator)
+            {
+                TelemetryClient = this.TelemetryClient
```
[the remark this answers is not in the document]
Comment (garypretty) [reply samples/csharp_dotnetcore/21.corebot-app-insights/Dialogs/DateResolverDialog.cs]: Removed.

File: samples/csharp_dotnetcore/21.corebot-app-insights/Dialogs/DateResolverDialog.cs [hunk 13]
```diff
@@ -6,21 +6,31 @@
...
+            });
             AddDialog(new WaterfallDialog(nameof(WaterfallDialog), new WaterfallStep[]
             {
                 InitialStepAsync,
                 FinalStepAsync,
-            }));
+            })
+            {
+                TelemetryClient = this.TelemetryClient
```
[the remark this answers is not in the document]
Comment (garypretty) [reply samples/csharp_dotnetcore/21.corebot-app-insights/Dialogs/DateResolverDialog.cs]: Removed.

File: samples/csharp_dotnetcore/21.corebot-app-insights/Dialogs/MainDialog.cs [hunk 14]
```diff
@@ -2,100 +2,152 @@
...
-            TelemetryClient = telemetryClient;
 
-            AddDialog(new TextPrompt(nameof(TextPrompt))
-            {
-                TelemetryClient = telemetryClient,
-            });
-            AddDialog(new BookingDialog(telemetryClient));
+            this.TelemetryClient = telemetryClient;
+
+            // The TelemetryClient property for each added dialog will be set to
```
Comment (WashingtonKayaker) [inline samples/csharp_dotnetcore/21.corebot-app-insights/Dialogs/MainDialog.cs]: Suggestions for comment:
* Move it above the line where you set the telemetryClient.
* Simplify the wording, perhaps something like:

>  // Set the telemetry client for this and all child dialogs.

Comment (garypretty) [reply samples/csharp_dotnetcore/21.corebot-app-insights/Dialogs/MainDialog.cs]: Done. Thanks for the suggestion.

File: samples/csharp_dotnetcore/21.corebot-app-insights/AdapterWithErrorHandler.cs [hunk 15]
```diff
@@ -3,28 +3,35 @@
...
             OnTurnError = async (turnContext, exception) =>
             {
                 // Log any leaked exception from the application.
                 logger.LogError(exception, $"[OnTurnError] unhandled error : {exception.Message}");
 
                 // Send a message to the user
-                await turnContext.SendActivityAsync("The bot encountered an error or bug.");
-                await turnContext.SendActivityAsync("To continue to run this bot, please fix the bot source code.");
+                var errorMessageText = "The bot encountered an error or bug.";
+                var errorMessage = MessageFactory.Text(errorMessageText, errorMessageText, InputHints.ExpectingInput);
```
[the remark this answers is not in the document]
Comment (garypretty) [reply samples/csharp_dotnetcore/21.corebot-app-insights/AdapterWithErrorHandler.cs]: Amended. Thanks for the reminder.

File: samples/csharp_dotnetcore/21.corebot-app-insights/FlightBookingRecognizer.cs [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+                    configuration["LuisAppId"],
+                    configuration["LuisAPIKey"],
+                    "https://" + configuration["LuisAPIHostName"]);
+
+                // Set the recognizer options depending on which endpoint version you want to use.
+                // More details can be found in https://docs.microsoft.com/en-gb/azure/cognitive-services/luis/luis-migration-api-v3
+                var recognizerOptions = new LuisRecognizerOptionsV3(luisApplication)
+                {
+                    TelemetryClient = telemetryClient,
+                };
```
Comment (garypretty) [inline samples/csharp_dotnetcore/21.corebot-app-insights/FlightBookingRecognizer.cs]: @WashingtonKayaker Just a note to say that I merged the latest changes in from master, which included an update for the LUIS recognizer here.

Comment (WashingtonKayaker) [reply samples/csharp_dotnetcore/21.corebot-app-insights/FlightBookingRecognizer.cs]: Thank you, I've updated the [docs](https://github.com/MicrosoftDocs/bot-docs-pr/pull/1909) to reflect this change.
